commands/create/nodepool/command.go

```diff
@@ -0,0 +1,357 @@
+	ScalingMin            int64
+	Scheme                string
+	Verbose               bool
+}
+
+type result struct {
+	nodePoolID            string
+	nodePoolName          string
+	availabilityZonesList []string
+}
+
+// collectArguments populates an arguments struct with values both from command flags,
+// from config, and potentially from built-in defaults.
+func collectArguments(positionalArgs []string) (Arguments, error) {
+	endpoint := config.Config.ChooseEndpoint(flags.CmdAPIEndpoint)
+	token := config.Config.ChooseToken(endpoint, flags.CmdToken)
+	scheme := config.Config.ChooseScheme(endpoint, flags.CmdToken)
+
+	zones := cmdAvailabilityZones
+	var err error
+	if zones != nil && len(zones) > 0 {
+		zones, err = expandZones(zones, endpoint, token)
+		if err != nil {
+			return Arguments{}, microerror.Mask(err)
+		}
+	}
+
+	return Arguments{
+		APIEndpoint:           endpoint,
+		AuthToken:             token,
+		AvailabilityZonesList: zones,
+		AvailabilityZonesNum:  cmdAvailabilityZonesNum,
+		ClusterID:             positionalArgs[0],
+		InstanceType:          flags.CmdWorkerAwsEc2InstanceType,
+		Name:                  cmdName,
+		ScalingMax:            flags.CmdWorkersMax,
+		ScalingMin:            flags.CmdWorkersMin,
+		Scheme:                scheme,
+		Verbose:               flags.CmdVerbose,
+	}, nil
+}
+
+// expandZones takes a list of alphabetical letters and returns a list of
+// availability zones. Example:
+//
+// ["a", "b"] -> ["eu-central-1a", "eu-central-1b"]
+//
+func expandZones(zones []string, endpoint, token string) ([]string, error) {
+	clientWrapper, err := client.NewWithConfig(endpoint, token)
+	if err != nil {
+		return []string{}, microerror.Mask(err)
+	}
+
+	info, err := clientWrapper.GetInfo(nil)
+	if err != nil {
+		return []string{}, microerror.Mask(err)
+	}
+
+	out := []string{}
+	for _, letter := range zones {
+		if len(letter) == 1 {
+			letter = info.Payload.General.Datacenter + strings.ToLower(letter)
+		}
+		out = append(out, letter)
+	}
+
+	return out, nil
+}
+
+func verifyPreconditions(args Arguments) error {
+	if config.Config.Token == "" && args.AuthToken == "" {
+		return microerror.Mask(errors.NotLoggedInError)
+	}
+
+	if args.ClusterID == "" {
+		return microerror.Mask(errors.ClusterIDMissingError)
+	}
+
+	// AZ flags plausibility
+	if len(args.AvailabilityZonesList) > 0 && args.AvailabilityZonesNum > 0 {
+		return microerror.Maskf(errors.ConflictingFlagsError, "the flags --availability-zones and --num-availability-zones cannot be combined.")
+	}
+
+	// Scaling flags plausibility
+	if args.ScalingMin > 0 && args.ScalingMax > 0 {
+		if args.ScalingMin > args.ScalingMax {
+			return microerror.Mask(errors.WorkersMinMaxInvalidError)
+		}
+	}
+
+	return nil
+}
+
+func printValidation(cmd *cobra.Command, positionalArgs []string) {
+	var err error
+	args, err := collectArguments(positionalArgs)
+	if err == nil {
+		err = verifyPreconditions(args)
+	}
+
+	if err == nil {
+		return
+	}
```

Comment:
This here is extremely complicated to understand for me now. It would be nice to maybe talk about it and make it better. It also looks like that `printResults` does basically the same but implements the logic differently.

Reply:
This is really how all gsctl commands are built currently.

`printValidation` is the pre-execution step which also triggers validation, then shows the result. If an error occured, it exits with a code 1.

`printResult` repeats the collectArguments step, but it does not have to expect and handle errors from that step, as those would already be handled before and it would not reach this stage.